src/JSON/Utils/Cursors.dfy

```diff
@@ -0,0 +1,329 @@
+include "../../BoundedInts.dfy"
+include "../../Wrappers.dfy"
+include "Views.dfy"
+include "Lexers.dfy"
+
+module {:options "-functionSyntax:4"} JSON.Utils.Cursors {
+  import opened BoundedInts
+  import opened Wrappers
+
+  import opened Vs = Views.Core
+  import opened Lx = Lexers.Core
+
+  datatype Split<+T> = SP(t: T, cs: FreshCursor) {
+    ghost predicate BytesSplitFrom?(cs0: Cursor, spec: T -> bytes) {
+      cs0.Bytes() == spec(t) + cs.Bytes()
+    }
+
+    ghost predicate SplitFrom?(cs0: Cursor, spec: T -> bytes) {
+      cs.SplitFrom?(cs0) && BytesSplitFrom?(cs0, spec)
+    }
+
+    ghost predicate StrictlySplitFrom?(cs0: Cursor, spec: T -> bytes) {
+      cs.StrictlySplitFrom?(cs0) && BytesSplitFrom?(cs0, spec)
+    }
+  }
+
+  // LATER: Make this a newtype and put members here instead of requiring `Valid?` everywhere
+  type Cursor = ps: Cursor_ | ps.Valid?
+    witness Cursor([], 0, 0, 0)
+  type FreshCursor = ps: Cursor | ps.BOF?
+    witness Cursor([], 0, 0, 0)
+
+  datatype CursorError<+R> =
+    | EOF
+    | ExpectingByte(expected: byte, b: opt_byte)
+    | ExpectingAnyByte(expected_sq: seq<byte>, b: opt_byte)
+    | OtherError(err: R)
+  {
+    function ToString(pr: R -> string) : string {
+      match this
+        case EOF => "Reached EOF"
+        case ExpectingByte(b0, b) =>
+          var c := if b > 0 then "'" + [b as char] + "'" else "EOF";
+          "Expecting '" + [b0 as char] + "', read " + c
+        case ExpectingAnyByte(bs0, b) =>
+          var c := if b > 0 then "'" + [b as char] + "'" else "EOF";
+          var c0s := seq(|bs0|, idx requires 0 <= idx < |bs0| => bs0[idx] as char);
+          "Expecting one of '" + c0s + "', read " + c
+        case OtherError(err) => pr(err)
+    }
+  }
+  type CursorResult<+R> = Result<Cursor, CursorError<R>>
+
+  datatype Cursor_ = Cursor(s: bytes, beg: uint32, point: uint32, end: uint32) {
```

Comment:
It will be good to generalize this code later to different integer sizes, but I would rather stick with machine int for the first version, for efficiency.

Reply:
kk, Then we have 2 somewhat reference line implementations as I see it,
the ESDKs and this one.
We should try and start working on a general solution that both these use cases could use :)